src/main/java/sparksoniq/jsoniq/item/DateTimeItem.java

```diff
@@ -0,0 +1,208 @@
+package sparksoniq.jsoniq.item;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Period;
+import org.joda.time.PeriodType;
+import org.rumbledb.api.Item;
+import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.exceptions.UnexpectedTypeException;
+import sparksoniq.jsoniq.compiler.translator.expr.operational.base.OperationalExpressionBase;
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.types.AtomicTypes;
+import sparksoniq.semantics.types.ItemType;
+import sparksoniq.semantics.types.ItemTypes;
+
+import java.util.regex.Pattern;
+
+
+public class DateTimeItem extends AtomicItem {
+
+    private static final String yearFrag = "((-)?(([1-9]\\d\\d(\\d)+)|(0\\d\\d\\d)))";
+    private static final String monthFrag = "((0[1-9])|(1[0-2]))";
+    private static final String dayFrag = "((0[1-9])|([1-2]\\d)|(3[0-1]))";
+    private static final String hourFrag = "(([0-1]\\d)|(2[0-3]))";
+    private static final String minuteFrag = "([0-5]\\d)";
+    private static final String secondFrag = "(([0-5]\\d)(\\.(\\d)+)?)";
+    private static final String endOfDayFrag = "(24:00:00(\\.(0)+)?)";
+    private static final String timezoneFrag = "(Z|([+\\-])(((0\\d|1[0-3]):" + minuteFrag + ")|(14:00)))";
+    private static final String dateTimeLexicalRep = yearFrag + "-" + monthFrag + "-" + dayFrag + "T" +
+            "((" + hourFrag + ":" + minuteFrag + ":" + secondFrag + ")|(" + endOfDayFrag + "))(" + timezoneFrag + ")?";
+    private static final Pattern dateTimePattern = Pattern.compile(dateTimeLexicalRep);
+
+    private static final long serialVersionUID = 1L;
+    private DateTime _value;
+    private boolean hasTimeZone = true;
+
+    public DateTimeItem() { super(); }
+
+    DateTimeItem(DateTime _value, boolean hasTimeZone) {
```

Comment:
I was indeed considering of changing all the itemtypes constructors from public to default, sooner or later. In the end, we should create Items through ItemFactory, that is in the same package, so there is no real need for the constructors to be public. Maybe one exception could be FunctionItem, that has quite a few params in the constructor and going through another function like _createFunctionItem(...)_ could not be worth it. Do you agree or there is a reason why constructors should be made public @CanBerker ?

Reply:
From an OOP perspective that sounds like a good idea. There is only one case that I'm curious about and it's the closures that get passed to Spark. I have noticed that CountClauseClosure and ReturnFlatMapClosure (through deserializeRowField method) makes calls to Itemfactory....create... It might be a micro optimization but as these may get applied to billions of items I wonder if calling a constructor directly, rather than a factory would bring about a better performance. If that would be the case then having public constructors could be beneficial. What do you think @ghislainfourny ?

Reply:
This is encapsulation at work here.

Going through the factory rather than directly calling constructors encapsulates the item classes, allowing for optimizations in the future such as pooling items. Indeed, the factory does not have to create new items every time a creation is requested -- for example, creating boolean and null items do not construct new objects every time but always return the same instances. In the future, similar optimizations can be done with other items as well when there are a lot of duplicates.

This is why the factory should be the way to create any items rather than directly invoking constructors. If the rest of the code only (design by contract) uses the factory to create items, and only navigates items with the Item class method, without using any insider knowledge about the subclasses, then all these optimizations will be possible without modifying code outside the item package.

Does it make sense?